src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHttpRequestOptionsExtensions.cs

```diff
@@ -0,0 +1,27 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+
+namespace System.Net.Http.Metrics;
+
+public static class MetricsHttpRequestOptionsExtensions
+{
+    // TODO: To eliminate the extra allocation of List<T> for common cases, define a custom collection that embeds callback instances up to a certain number.
+    private static readonly HttpRequestOptionsKey<List<Action<HttpMetricsEnrichmentContext>>> s_callbackCollectionKey = new("MetricsEnrichmentCallbackCollection");
+
+    public static void AddMetricsEnrichmentCallback(this HttpRequestOptions options, Action<HttpMetricsEnrichmentContext> callback)
+    {
+        if (!options.TryGetValue(s_callbackCollectionKey, out List<Action<HttpMetricsEnrichmentContext>>? callbackCollection))
+        {
+            callbackCollection = new List<Action<HttpMetricsEnrichmentContext>>();
```

Comment:
Is the intent to pool this?  Otherwise, it's another non-trivial set of allocations (the list, the wrapped array, and potentially more arrays if growth is needed) on every request?

Reply:
Unlike with `HttpMetricsEnrichmentContext`, where we can assume that the callbacks are running on the same thread, the delegate list is trickier to pool. Instead, I was planning to implement a custom collection which would embed a few callbacks, so it would be only one allocation in most cases.

Reply:
How many callbacks are expected in the common cases?

Reply:
This is still something to figure out.

Reply:
Any concrete scenario I know if is either 0 or 1. I can come up with something hypothetical for more than 1 but it feels pretty unlikely at the moment.

Reply:
If we are anticipating a perf advantage where 1 callback doesn't allocate but more than 1 does allocate, I bet that is enough incentive that Martin will ensure all their enrichment behavior shares a single callback. You are right though, I wasn't considering how they might want to adjust their implementation based on some of the middleware changes we were proposing. If there was neglible perf difference to use a 2nd callback, they might do that.